README.md
# issue-expander

*Expand GitHub issue references into Markdown links*

[![PyPI](https://img.shields.io/pypi/v/issue-expander?color=green&logo=python&logoColor=white)](https://pypi.org/project/issue-expander/)
![Coverage](https://img.shields.io/endpoint?url=https://gist.githubusercontent.com/adamwolf/4537e853375d0289662b6c7741571cb0/raw/covbadge.json)
![PyPI - Python Version](https://img.shields.io/pypi/pyversions/issue_expander)

```
$ echo 'rust-lang/rust#106827' | issue-expander -
[Update LLVM to 15.0.7 #106827](https://github.com/rust-lang/rust/pull/106827)

$ echo '#106827' | issue-expander --default-source 'rust-lang/rust'
[Update LLVM to 15.0.7 #106827](https://github.com/rust-lang/rust/pull/106827)
```

This project is not yet stable in any way and makes absolutely no guarantees about behavior
(between releases, or otherwise).

## Usage

<!-- [[[cog
import cog
from issue_expander import expander
from click.testing import CliRunner
runner = CliRunner()
result = runner.invoke(expander.cli, ["--help"])
help = result.output.replace("Usage: cli", "Usage: issue-expander")
cog.out(
    "```\n{}\n```".format(help)
)
]]] -->
```
Usage: issue-expander [OPTIONS] FILE

  Turn references like "foo/bar#123" into Markdown links, like

  "[Prevent side fumbling #123](https://github.com/foo/bar/pull/123)"

  issue-expander works for references to issues and to pull requests.

  References are only expanded if they are found at GitHub.  To expand
  references from private repositories, you'll need to pass your GitHub token.
  This can be done via environment variables or via command line options.

  To interpret references like `#1138` as `adamwolf/issue-expander#1138`,
  specify defaults using `--default-source`.

Options:
  --default-source USER/REPO  Use USER/REPO when not specified in issue
                              reference. (Example: "adamwolf/issue-expander")
  -p, --github-token TOKEN    GitHub token for looking up issue references. You
                              may want to use the environment variable
                              ISSUE_EXPANDER_GITHUB_TOKEN instead.
  --version                   Show the version and exit.
  --help                      Show this message and exit.

```
<!-- [[[end]]] -->

## Installation

If you use Homebrew, you can install issue-expander like this:
<!-- [[[cog
import cog
with open("_docs/install-via-homebrew.txt") as f:
    cog.out("```\n")
    for line in f:
        cog.out(line)
    cog.out("```\n")
]]] -->
```
brew tap adamwolf/issue-expander
brew install issue-expander
```
<!-- [[[end]]] -->

Otherwise, I recommend installing with [pipx](https://pypa.github.io/pipx/). Once you've got pipx installed, you can install issue-expander like this:

<!-- [[[cog
import cog
with open("_docs/install-via-pipx.txt") as f:
    cog.out("```\n")
    for line in f:
        cog.out(line)
    cog.out("```\n")
]]] -->
```
pipx install issue-expander
```
<!-- [[[end]]] -->

and pipx installs issue-expander into its own little area, just for you, where you don't have to worry about it messing with any system-wide or even user-wide Python dependencies.

If you want to generate a binary that can run issue-expander without needing Python to be installed, see [BUILDING.md](BUILDING.md).

If you want to do something different, issue-expander is available on PyPI. Go hogwild.

## Development

There are a lot of ways to set up a development environment for a Python application.  This is one of them.

Before you follow these steps, make sure Python 3 is installed.

1. Clone the issue-expander repository.
<!-- [[[cog
import cog
with open("_docs/devclone.txt") as f:
    cog.out("```\n")
    for line in f:
        cog.out(line)
    cog.out("```\n")
]]] -->
```
git clone https://github.com/adamwolf/issue-expander.git
```
<!-- [[[end]]] -->

2. Create a virtual environment and activate it.

<!-- [[[cog
import cog
with open("_docs/devvenv.txt") as f:
    cog.out("```\n")
    for line in f:
        cog.out(line)
    cog.out("```\n")
]]] -->
```
cd issue-expander
python3 -m venv venv
source venv/bin/activate
```
<!-- [[[end]]] -->

3. Install the package in editable mode (with development extras).
<!-- [[[cog
import cog
with open("_docs/devinstall.txt") as f:
    cog.out("```\n")
    for line in f:
        cog.out(line)
    cog.out("```\n")
]]] -->
```
pip install -e .[dev]
```
<!-- [[[end]]] -->

4. Make sure you can run the tests.
<!-- [[[cog
import cog
with open("_docs/devtest.txt") as f:
    cog.out("```\n")
    for line in f:
        cog.out(line)
    cog.out("```\n")
]]] -->
```
pytest
```
<!-- [[[end]]] -->

5. Make sure you can run the linter.
<!-- [[[cog
import cog
with open("_docs/devlint.txt") as f:
    cog.out("```\n")
    for line in f:
        cog.out(line)
    cog.out("```\n")
]]] -->
```
pre-commit run --all-files
```
<!-- [[[end]]] -->

At this point, you should be able to make changes to the code.

# Contributing

See [CONTRIBUTING.md](CONTRIBUTING.md).


RELEASES.md
# Releases

## Preparation

Bump the version in `pyproject.toml` by removing the `.devx`.  Update the test!

Use `scriv collect` to create an initial changelog.  Review it.

Make a branch. Make a commit with the `changelog.d/` removals, the `CHANGELOG.md` changes,
the `pyproject.toml` version change...

After the PR is merged, pull, and make a tag with `git tag vx.y.z`.

Push the tag with `git push --tags`.

## Build PyPI packages and wheels

(It would be great to build more wheels here!)

```
pip install build twine
rm -rf build dist
python -m build .
twine check dist/*`
```

Do whatever testing you need.  If you did anything weird or to metadata, you can push to Test PyPI with `twine upload -r test dist/*`, otherwise:

`twine upload -r issue-expander dist/*`


## Homebrew tap

Update homebrew-issue-expander with `brew bump-version-pr issue-expander --version x.y.z`

Once the pull request passes the builds, add a `pr-pull` tag to continue the automation.

## Get ready for more development

Make a new branch.  Increment the version number and add a `.dev0`.  (Update the test!) Make a pull request and merge it.

## GitHub release
scriv will create a GitHub release based on the latest tag and entry in the CHANGELOG.md file.

```
scriv github-release --dry-run
export GITHUB_TOKEN=password1
scriv github-release
```


tests/conftest.py
import pytest

from issue_expander.expander import getIssue


@pytest.fixture(autouse=True)
def clear_issue_cache():
    getIssue.cache_clear()


tests/test_cli.py
"""
Use the CLI directly for some high-level integration tests.
"""
import pytest
from click.testing import CliRunner

import issue_expander
from issue_expander.expander import cli


def test_file(monkeypatch):
    """Use a file as input."""

    def mockIssue(owner, repository, number, token):
        if owner == "adamwolf" and repository == "geewhiz" and number == "101" and token is None:
            return {"html_url": "https://example.com/pulls/106970", "title": "Foobar the Frobnitz"}
        else:
            raise ValueError("Unexpected request")

    monkeypatch.setattr(issue_expander.expander, "getIssue", mockIssue)

    runner = CliRunner()
    with runner.isolated_filesystem():
        with open("hello.txt", "w") as f:
            f.write("adamwolf/geewhiz#101")

        result = runner.invoke(cli, ["hello.txt"])
        assert result.output == "[Foobar the Frobnitz #101](https://example.com/pulls/106970)"
        assert result.exit_code == 0


def test_stdin(monkeypatch):
    """Read input from stdin if given - as the input file."""

    def mockIssue(owner, repository, number, token):
        if owner == "adamwolf" and repository == "geewhiz" and number == "101" and token is None:
            return {"html_url": "https://example.com/pulls/106970", "title": "Foobar the Frobnitz"}
        else:
            raise ValueError("Unexpected request")

    monkeypatch.setattr(issue_expander.expander, "getIssue", mockIssue)

    runner = CliRunner()
    result = runner.invoke(cli, ["-"], input="adamwolf/geewhiz#101")
    assert result.output == "[Foobar the Frobnitz #101](https://example.com/pulls/106970)"
    assert result.exit_code == 0


def test_gh_expansion(monkeypatch):
    """Expand GH-123 style references."""

    def mockIssue(owner, repository, number, token):
        if owner == "adamwolf" and repository == "geewhiz" and number == "123" and token is None:
            return {"html_url": "https://example.com/pull/123", "title": "Hoist the Mainsail!"}
        else:
            raise ValueError("Unexpected request")

    monkeypatch.setattr(issue_expander.expander, "getIssue", mockIssue)

    runner = CliRunner()
    result = runner.invoke(cli, ["--default-source", "adamwolf/geewhiz", "-"], input="GH-123")
    assert result.output == "[Hoist the Mainsail! #123](https://example.com/pull/123)"
    assert result.exit_code == 0


def test_url_expansion(monkeypatch):
    """Expand url references."""

    def mockIssue(owner, repository, number, token):
        if owner == "adamwolf" and repository == "geewhiz" and number == "123" and token is None:
            return {"html_url": "https://example.com/pull/123", "title": "Hoist the Mainsail!"}
        else:
            raise ValueError("Unexpected request")

    monkeypatch.setattr(issue_expander.expander, "getIssue", mockIssue)

    runner = CliRunner()
    result = runner.invoke(cli, ["-"], input="https://github.com/adamwolf/geewhiz/pull/123")
    assert result.output == "[Hoist the Mainsail! #123](https://example.com/pull/123)"
    assert result.exit_code == 0


def test_url_expansion_with_mismatched_urls(monkeypatch):
    """Expand url references when the url is for an issue and the item is a PR."""

    def mockIssue(owner, repository, number, token):
        if owner == "adamwolf" and repository == "geewhiz" and number == "123" and token is None:
            return {"html_url": "https://example.com/pulls/123", "title": "Hoist the Mainsail!"}
        else:
            raise ValueError("Unexpected request")

    monkeypatch.setattr(issue_expander.expander, "getIssue", mockIssue)

    runner = CliRunner()
    result = runner.invoke(cli, ["-"], input="https://github.com/adamwolf/geewhiz/issues/123")
    assert result.output == "[Hoist the Mainsail! #123](https://example.com/pulls/123)"
    assert result.exit_code == 0


def test_default_source(monkeypatch):
    """Use a default owner/repository when specified as an option."""

    def mockIssue(owner, repository, number, token):
        if owner == "adamwolf" and repository == "geewhiz" and number == "101" and token is None:
            return {"html_url": "https://example.com/pulls/106970", "title": "Foobar the Frobnitz"}
        else:
            raise ValueError("Unexpected request")

    monkeypatch.setattr(issue_expander.expander, "getIssue", mockIssue)

    runner = CliRunner()
    result = runner.invoke(cli, ["--default-source", "adamwolf/geewhiz", "-"], input="#101")
    assert result.output == "[Foobar the Frobnitz #101](https://example.com/pulls/106970)"
    assert result.exit_code == 0


bad_default_sources = [
    "",
    "https://www.google.com",
    "      adamwolf/geewhiz",
    "adamwolf     /      geewhiz",
    "adamwolf",
    "adamwolf/",
    "/geewhiz",
    "adamwolf/geewhiz/",
    "adamwolf//geewhiz",
    "/" "//" "a/b/c" " hello world" "01189998819991197253",
]


@pytest.mark.parametrize("source", bad_default_sources)
def test_malformed_default_sources(source):
    """Malformed default sources are rejected with a usage error."""
    runner = CliRunner()
    result = runner.invoke(cli, ["--default-source", source, "-"], input="#101")
    assert result.exit_code == 1
    assert "Error: default source must be in the format 'owner/repository'\n" in result.output


def test_default_source_but_not_needed(monkeypatch):
    """Don't override a specified owner/repository even if a default is specified"""

    def mockIssue(owner, repository, number, token):
        if owner == "adamwolf" and repository == "geewhiz" and number == "101" and token is None:
            return {"html_url": "https://example.com/pulls/106970", "title": "Foobar the Frobnitz"}
        else:
            raise ValueError("Unexpected request")

    monkeypatch.setattr(issue_expander.expander, "getIssue", mockIssue)

    runner = CliRunner()
    result = runner.invoke(
        cli,
        ["--default-source", "adamwolf/defaultdefault", "-"],
        input="adamwolf/geewhiz#101",
    )
    assert result.output == "[Foobar the Frobnitz #101](https://example.com/pulls/106970)"
    assert result.exit_code == 0


def test_two_expansions_in_one_line(monkeypatch):
    """Test that we can expand multiple issues in one line"""

    def mockIssue(owner, repository, number, token):
        if owner == "adamwolf" and repository == "geewhiz" and number == "101" and token is None:
            return {"html_url": "https://example.com/pulls/101", "title": "Foobar the Frobnitz"}
        elif owner == "adamwolf" and repository == "geewhiz" and number == "102" and token is None:
            return {"html_url": "https://example.com/pulls/102", "title": "HOTFIX: fix the frobnitz!"}
        else:
            raise ValueError("Unexpected request")

    monkeypatch.setattr(issue_expander.expander, "getIssue", mockIssue)

    runner = CliRunner()
    result = runner.invoke(cli, ["-"], input="adamwolf/geewhiz#101 adamwolf/geewhiz#102")
    assert result.output == (
        "[Foobar the Frobnitz #101](https://example.com/pulls/101) "
        "[HOTFIX: fix the frobnitz! #102](https://example.com/pulls/102)"
    )


def test_multiline_expansions_over_stdin(monkeypatch):
    """Expand issues on each line when given multiple lines through stdin"""

    def mockIssue(owner, repository, number, token):
        if owner == "adamwolf" and repository == "geewhiz" and number == "101" and token is None:
            return {"html_url": "https://example.com/pulls/101", "title": "Foobar the Frobnitz"}
        elif owner == "adamwolf" and repository == "geewhiz" and number == "102" and token is None:
            return {"html_url": "https://example.com/pulls/102", "title": "HOTFIX: fix the frobnitz!"}
        else:
            raise ValueError("Unexpected request")

    monkeypatch.setattr(issue_expander.expander, "getIssue", mockIssue)

    runner = CliRunner()
    result = runner.invoke(cli, ["-"], input="adamwolf/geewhiz#101\nhello world\nadamwolf/geewhiz#102")
    assert result.output == (
        "[Foobar the Frobnitz #101](https://example.com/pulls/101)\n"
        "hello world\n"
        "[HOTFIX: fix the frobnitz! #102](https://example.com/pulls/102)"
    )
    assert result.exit_code == 0


def test_multiline_expansions_over_file(monkeypatch):
    """Expand issues on each line when given multiple lines in a file"""

    def mockIssue(owner, repository, number, token):
        if owner == "adamwolf" and repository == "geewhiz" and number == "101" and token is None:
            return {"html_url": "https://example.com/pulls/101", "title": "Foobar the Frobnitz"}
        elif owner == "adamwolf" and repository == "geewhiz" and number == "102" and token is None:
            return {"html_url": "https://example.com/pulls/102", "title": "HOTFIX: fix the frobnitz!"}
        else:
            raise ValueError("Unexpected request")

    monkeypatch.setattr(issue_expander.expander, "getIssue", mockIssue)

    runner = CliRunner()
    with runner.isolated_filesystem():
        with open("hello.txt", "w") as f:
            f.write("adamwolf/geewhiz#101\nhello world\nadamwolf/geewhiz#102")

        result = runner.invoke(cli, ["hello.txt"])
        assert result.output == (
            "[Foobar the Frobnitz #101](https://example.com/pulls/101)\n"
            "hello world\n"
            "[HOTFIX: fix the frobnitz! #102](https://example.com/pulls/102)"
        )
        assert result.exit_code == 0


def test_version():
    """--version prints the version"""
    runner = CliRunner()
    result = runner.invoke(cli, ["--version"])
    assert result.output == "cli, version 0.1.13.dev0\n"
    assert result.exit_code == 0


def test_help():
    """--help prints a helpful message"""
    runner = CliRunner()
    result = runner.invoke(cli, ["--help"])
    assert result.exit_code == 0
    assert "Usage:" in result.output
    assert "Options:" in result.output
    assert 'Turn references like "foo/bar#123" into Markdown links' in result.output


def test_cli_auth(monkeypatch):
    """Credentials can be passed on the command line"""

    def mockIssue(owner, repository, number, token):
        if owner == "foo" and repository == "privaterepo" and number == "555" and token == "password1":
            return {"html_url": "https://example.com/pulls/555", "title": "Sshhh!"}
        else:
            raise ValueError("Unexpected request")

    monkeypatch.setattr(issue_expander.expander, "getIssue", mockIssue)

    runner = CliRunner()
    result = runner.invoke(
        cli,
        ["--github-token", "password1", "-"],
        input="foo/privaterepo#555",
    )
    assert result.output == "[Sshhh! #555](https://example.com/pulls/555)"
    assert result.exit_code == 0


def test_404s_dont_expand(monkeypatch):
    """If we get a 404 when looking up an issue, don't expand it."""

    def mockIssue(owner, repository, number, token):
        if owner == "adamwolf" and repository == "geewhiz" and number == "101" and token is None:
            return None
        else:
            raise ValueError("Unexpected request")

    monkeypatch.setattr(issue_expander.expander, "getIssue", mockIssue)

    runner = CliRunner()
    with runner.isolated_filesystem():
        with open("hello.txt", "w") as f:
            f.write("adamwolf/geewhiz#101")

        result = runner.invoke(cli, ["hello.txt"])
        assert result.output == "adamwolf/geewhiz#101"
        assert result.exit_code == 0


tests/test_getIssue.py
import io
import json
from unittest.mock import patch
from urllib.error import HTTPError

import issue_expander
from issue_expander.expander import getIssue


def test_getIssue_with_404(monkeypatch):
    """getIssue should return None if the issue is not found"""

    def mock_urlopen(request, context=None):
        raise HTTPError("http://example.com", 404, "Not Found", {}, io.BytesIO(b""))

    monkeypatch.setattr(issue_expander.expander, "urlopen", mock_urlopen)
    assert getIssue("adamwolf", "issue-expander", "1", None) is None


def test_getIssue_puts_token_in_headers():
    """getIssue should put the token in the headers as Authorization"""

    with patch(
        "issue_expander.expander.urlopen",
        return_value=io.BytesIO(json.dumps({"html_url": "foo", "title": "bar"}).encode("utf-8")),
    ) as mock_urlopen:
        assert getIssue("adamwolf", "issue-expander", "1", "frobnitz") == {
            "html_url": "foo",
            "title": "bar",
        }
        mock_urlopen.assert_called_once
        request_headers = mock_urlopen.call_args[0][0].headers
        assert request_headers["Authorization"] == "Bearer: frobnitz"


def test_getIssue_with_403_without_token(monkeypatch, capsys):
    """getIssue should return None but print an error if GitHub rate limits us.

    If we didn't pass a token it, it should tell us to try again with a token."""

    def mock_urlopen(request, context=None):
        raise HTTPError("http://example.com", 403, "Rate limit exceeded.", {}, io.BytesIO(b""))

    monkeypatch.setattr(issue_expander.expander, "urlopen", mock_urlopen)
    assert getIssue("adamwolf", "issue-expander", "1", None) is None
    captured = capsys.readouterr()
    assert (
        captured.err
        == "Unable to look up issue due to rate limit error. Try providing a token with --token.\n"
    )


def test_getIssue_with_403_with_token(monkeypatch, capsys):
    """getIssue should return None but print an error if GitHub rate limits us.

    If we passed a token in, it shouldn't tell us to try again with a token."""

    def mock_urlopen(request, context=None):
        raise HTTPError("http://example.com", 403, "Rate limit exceeded.", {}, io.BytesIO(b"foo   "))

    monkeypatch.setattr(issue_expander.expander, "urlopen", mock_urlopen)
    assert getIssue("adamwolf", "issue-expander", "1", "imatoken") is None
    captured = capsys.readouterr()
    assert captured.err == "Unable to look up issue due to rate limit error.\n"


def test_getIssue_with_malformed_json(monkeypatch, capsys):
    """getIssue should return None and print an error if the JSON is malformed"""

    def mock_urlopen(request, context=None):
        raise json.JSONDecodeError("msg", "doc", 0)

    monkeypatch.setattr(issue_expander.expander, "urlopen", mock_urlopen)
    assert getIssue("adamwolf", "issue-expander", "1", None) is None

    captured = capsys.readouterr()

    assert (
        captured.err
        == "Unable to parse response from https://api.github.com/repos/adamwolf/issue-expander/issues/1\n"
    )


def test_getIssue_happy_path():
    """getIssue should return the issue if it's found"""

    with patch(
        "issue_expander.expander.urlopen",
        return_value=io.BytesIO(json.dumps({"html_url": "foo", "title": "bar"}).encode("utf-8")),
    ) as mock_urlopen:
        assert getIssue("adamwolf", "issue-expander", "1", None) == {
            "html_url": "foo",
            "title": "bar",
        }
        mock_urlopen.assert_called_once


src/issue_expander/issue_expander.py
import re
import sys
from functools import lru_cache, partial

import click
import requests

# These regexes are really particular, at the moment..
regexes = [
    # like https://github.com/foo/bar/issues/123
    r"https://github.com/(?P<group>[a-zA-Z0-9.-]+)/(?P<repository>[a-zA-Z0-9.-]+)/(issues|pull)/(?P<number>\d+)",
    r"(?<![a-zA-Z0-9#-])#(?P<number>\d+)(?![a-zA-Z0-9#])",  # like #123
    r"\bGH-(?P<number>\d+)",  # like GH-123
    r"(?P<group>[a-zA-Z0-9.-]+)/(?P<repository>[a-zA-Z0-9.-]+)#(?P<number>\d+)",  # like foo/bar#123
]


@lru_cache()
def getIssue(group: str, repository: str, number: int, username: [str], token: [str]) -> [dict]:
    # What to do with bad credentials?

    url = f"https://api.github.com/repos/{group}/{repository}/issues/{number}"
    if username and token:
        auth = (username, token)
    else:
        auth = None
    req = requests.get(url, auth=auth)
    if not req.ok:
        return None
    else:
        return req.json()


def substituteMatch(match, default_group, default_repository, username, token):
    if "group" in match.groupdict() and "repository" in match.groupdict():
        group = match.group("group")
        repository = match.group("repository")
    else:
        group = default_group
        repository = default_repository

    number = match.group("number")

    if group and repository:
        issue = getIssue(group, repository, number, username, token)
        if issue:
            html_url = issue["html_url"]
            title = issue["title"]
            # TODO do we need to escape [] or anything?
            return f"[{title} #{number}]({html_url})"
    # if we didn't find an issue, return the original text, so we don't change anything
    return match.group(0)


def expandRefsToMarkdown(
    text: str,
    username: object = None,
    token: object = None,
    default_group: object = None,
    default_repository: object = None,
) -> str:

    substituter = partial(
        substituteMatch,
        default_group=default_group,
        default_repository=default_repository,
        username=username,
        token=token,
    )
    # now substituter can be called with just one argument, the match

    out = text
    # make all the substitutions for this regex at one time,
    # then check that output string with the next regex

    for regex in regexes:
        out = re.sub(regex, substituter, out)
    return out


@click.command()
@click.option(
    "--default-source",
    metavar="USER/REPO",
    help='Use USER/REPO when not specified in issue reference. (Example: "adamwolf/issue-expander")',
)
@click.option(
    "-u",
    "--github-username",
    envvar="ISSUE_EXPANDER_GITHUB_USERNAME",
    metavar="USERNAME",
    help="GitHub username for looking up issue references. You can use the environment variable "
    "ISSUE_EXPANDER_GITHUB_USERNAME.",
)
@click.option(
    "-p",
    "--github-token",
    envvar="ISSUE_EXPANDER_GITHUB_TOKEN",
    metavar="TOKEN",
    help="GitHub token for looking up issue references. You may want to use the environment variable "
    "ISSUE_EXPANDER_GITHUB_TOKEN instead.",
)
@click.version_option()
@click.argument("input", metavar="FILE", type=click.File("r"))
def cli(input, github_username=None, github_token=None, default_source=None):
    """Turn references like "foo/bar#123" into Markdown links, like

    "[Prevent side fumbling #123](https://github.com/foo/bar/pull/123)"

    issue-expander works for references to issues and to pull requests.

    References are only expanded if they are found at GitHub.  To expand references from private
    repositories, you'll need to pass your GitHub username and token.  This can be done via
    environment variables or via command line options.

    To interpret references like `#1138` as `adamwolf/issue-expander#1138`,
    specify defaults using `--default-source`.
    """

    out = []

    default_group = None
    default_repository = None

    if default_source is not None:
        # default_group, default_repository = default_source.split("/")
        # get group/repository from default_source using a regex
        match = re.fullmatch(r"(?P<group>[a-zA-Z0-9.-]+)/(?P<repository>[a-zA-Z0-9.-]+)", default_source)
        if match:
            default_group = match.group("group")
            default_repository = match.group("repository")

        if not (default_group and default_repository):
            print("Error: default source must be in the format 'group/repository'", file=sys.stderr)
            sys.exit(1)

    for line in input:
        out = expandRefsToMarkdown(line, github_username, github_token, default_group, default_repository)
        print(out, end="")


src/issue_expander/expander.py
import json
import re
import ssl
import sys
from functools import lru_cache, partial
from urllib.error import HTTPError
from urllib.request import Request, urlopen

import certifi
import click

# These regexes are really particular, at the moment...
# When adding regexes, take care to only match if you should--this may mean negative look-ahead and -behind assertions
# and name the group number, owner, and repository, but append _ and a unique suffix.

# Python's re module doesn't support regexes containing multiple named groups with the same name.

# These regexes are later combined into a megaregex.
named_regexes = [
    # like https://github.com/foo/bar/issues/123
    r"https://github.com/(?P<owner_x>[a-zA-Z0-9.-]+)/(?P<repository_x>[a-zA-Z0-9.-]+)/(issues|pull)/(?P<number_x>\d+)",
    r"(?<![a-zA-Z0-9#-])#(?P<number_y>\d+)(?![a-zA-Z0-9#])",  # like #123
    r"\bGH-(?P<number_z>\d+)",  # like GH-123
    r"(?P<owner_n>[a-zA-Z0-9.-]+)/(?P<repository_n>[a-zA-Z0-9.-]+)#(?P<number_n>\d+)",  # like foo/bar#123
]

megaregex = re.compile("|".join(named_regexes))


@lru_cache()
def getIssue(owner: str, repository: str, number: int, token: [str]) -> [dict]:
    # What to do with bad credentials?

    url = f"https://api.github.com/repos/{owner}/{repository}/issues/{number}"

    headers = {"Accept": "application/json"}
    if token:
        headers["Authorization"] = f"Bearer: {token}"

    request = Request(url, headers=headers)

    certifi_context = ssl.create_default_context(cafile=certifi.where())

    try:
        with urlopen(request, context=certifi_context) as response:
            j = json.load(response)
            return j
    except HTTPError as e:
        if e.code == 403:  # "rate limit exceeded"
            message = "Unable to look up issue due to rate limit error."
            if not token:
                message += " Try providing a token with --token."
            print(message, file=sys.stderr)
    except json.JSONDecodeError:
        print(f"Unable to parse response from {url}", file=sys.stderr)
        return None


def substituteMatch(match, default_owner, default_repository, token):
    owner = None
    repository = None
    number = None

    for name, value in match.groupdict().items():
        if name.startswith("number_") and value is not None:
            number = value
        elif name.startswith("owner_") and value is not None:
            owner = value
        elif name.startswith("repository_") and value is not None:
            repository = value

    if not (owner and repository):
        owner = default_owner
        repository = default_repository

    if owner and repository:
        issue = getIssue(owner, repository, number, token)
        if issue:
            html_url = issue["html_url"]
            title = issue["title"]
            # TODO do we need to escape [] or anything?
            return f"[{title} #{number}]({html_url})"
    # if we didn't find an issue, return the original text, so we don't change anything
    return match.group(0)


def expandRefsToMarkdown(
    text: str,
    token: object = None,
    default_owner: object = None,
    default_repository: object = None,
) -> str:
    substituter = partial(
        substituteMatch,
        default_owner=default_owner,
        default_repository=default_repository,
        token=token,
    )
    # now substituter can be called with just one argument, the match

    out = text
    # make all the substitutions for this regex at one time,
    # then check that output string with the next regex

    out = re.sub(megaregex, substituter, out)
    return out


@click.command()
@click.option(
    "--default-source",
    metavar="USER/REPO",
    help='Use USER/REPO when not specified in issue reference. (Example: "adamwolf/issue-expander")',
)
@click.option(
    "-p",
    "--github-token",
    envvar="ISSUE_EXPANDER_GITHUB_TOKEN",
    metavar="TOKEN",
    help="GitHub token for looking up issue references. You may want to use the environment variable "
    "ISSUE_EXPANDER_GITHUB_TOKEN instead.",
)
@click.version_option()
@click.argument("input", metavar="FILE", type=click.File("r"))
def cli(input, github_token=None, default_source=None):
    """Turn references like "foo/bar#123" into Markdown links, like

    "[Prevent side fumbling #123](https://github.com/foo/bar/pull/123)"

    issue-expander works for references to issues and to pull requests.

    References are only expanded if they are found at GitHub.  To expand references from private
    repositories, you'll need to pass your GitHub token.  This can be done via environment
    variables or via command line options.

    To interpret references like `#1138` as `adamwolf/issue-expander#1138`,
    specify defaults using `--default-source`.
    """

    out = []

    default_owner = None
    default_repository = None

    if default_source is not None:
        # get owner/repository from default_source using a regex
        match = re.fullmatch(r"(?P<owner>[a-zA-Z0-9.-]+)/(?P<repository>[a-zA-Z0-9.-]+)", default_source)
        if match:
            default_owner = match.group("owner")
            default_repository = match.group("repository")

        if not (default_owner and default_repository):
            print("Error: default source must be in the format 'owner/repository'", file=sys.stderr)
            sys.exit(1)

    for line in input:
        out = expandRefsToMarkdown(line, github_token, default_owner, default_repository)
        print(out, end="")
